gdal/frmts/fits/fitsdataset.cpp

```diff
@@ -599,6 +720,684 @@ GDALDataset *FITSDataset::Create( const char* pszFilename,
+            *pbSuccess = TRUE;
+
+        return dfNoDataValue;
+    }
+
+    if( poDS->bNoDataSet )
+    {
+        if( pbSuccess )
+            *pbSuccess = TRUE;
+
+        return poDS->dfNoDataValue;
+    }
+
+    return GDALPamRasterBand::GetNoDataValue( pbSuccess );
+}
+
+/************************************************************************/
+/*                           SetNoDataValue()                           */
+/************************************************************************/
+
+CPLErr FITSRasterBand::SetNoDataValue( double dfNoData )
+
+{
+    poDS->LoadGeoreferencingAndPamIfNeeded();
+
+    if( poDS->bNoDataSet && poDS->dfNoDataValue == dfNoData )
+    {
+        bNoDataSet = true;
+        dfNoDataValue = dfNoData;
+        return CE_None;
+    }
+
+    poDS->bNoDataSet = true;
+    poDS->dfNoDataValue = dfNoData;
+
+    poDS->bNoDataChanged = true;
+
+    bNoDataSet = true;
+    dfNoDataValue = dfNoData;
+    return CE_None;
+}
+
+/************************************************************************/
+/*                        DeleteNoDataValue()                           */
+/************************************************************************/
+
+CPLErr FITSRasterBand::DeleteNoDataValue()
+
+{
+    poDS->LoadGeoreferencingAndPamIfNeeded();
+
+    if( !poDS->bNoDataSet )
+        return CE_None;
+
+    poDS->bNoDataSet = false;
+    poDS->dfNoDataValue = -9999.0;
+
+    poDS->bNoDataChanged = true;
+
+    bNoDataSet = false;
+    dfNoDataValue = -9999.0;
+    return CE_None;
+}
+
+/************************************************************************/
+/*                     LoadGeoreferencingAndPamIfNeeded()               */
+/************************************************************************/
+
+void FITSDataset::LoadGeoreferencingAndPamIfNeeded()
+
+{
+    int status = 0;
+    int bitpix;
+    double dfScale, dfOffset;
+    double crpix1, crpix2, crval1, crval2, cdelt1, cdelt2, pc[4], cd[4];
+    double aRadius, cRadius, invFlattening = 0.0;
+    char target[16], ctype[9];
+    char pszGeogName[16], pszDatumName[16], projName[54];
+
+    const double PI = std::atan(1.0)*4;
+    const double DEG2RAD = PI / 180.;
+
+    if( !m_bReadGeoTransform && !m_bLoadPam )
+        return;
+
+/*
+    IdentifyAuthorizedGeoreferencingSources();
+*/
+/* -------------------------------------------------------------------- */
+/*      Get the transform or gcps from the FITS file.                */
+/* -------------------------------------------------------------------- */
+    if( m_bReadGeoTransform )
+    {
+        m_bReadGeoTransform = false;
+
+        OGRSpatialReference oSRS;
+
+        fits_read_key(hFITS, TSTRING, "OBJECT", target, nullptr, &status);
+        if( status )
+        {
+            strncpy(target, "Undefined", 10);
+            CPLError(CE_Failure, CPLE_AppDefined,
+                 "OBJECT keyword is missing");
+            status = 0;
+        }
+
+        strncpy(pszGeogName, "GCS_",5);
```

Comment:
risk of overflow. Use std::string to avoid that risky and tricky string length computation

Reply:
Fixed in 41699ae24b983381a1a63715fbbae71c442059d8